designs/0004-sparse-matrices.md

```diff
@@ -0,0 +1,387 @@
+- Sparse matrix-matrix multiplication (returns a sparse matrix that likely has a different sparsity structure)
+- Sparse matrix-dense matrix multiplication (should return a dense matrix)
+- Sparse matrix-diagonal matrix multiplication on the left and right (returns sparse matrix, same sparsity)
+- Sparse inner product and quadratic form (returns scalars)
+- Operations to move from sparse to dense matrices and vice versa.
+- Fill-reducing reorderings
+- A sparse Cholesky for a matrix of doubles.
+- The computation of the log-determinant as the product of Cholesky diagonals
+- Sparse linear solves and sparse triangular solves (for sampling from the marginalized out parameters in the generated quantities block)
+- Sparse algorithms to compute the required elements of the inverse.
+- An implementation of the reverse mode derivative of the log determinant
+- Reverse-mode derivative of a Cholesky decomposition of a sparse matrix.
+  - This is the same algorithm that was initially used in the dense case before Rob implemented Iain Murray’s blocked version. But it will need to be implemented in a “data structure” aware way
+- Specialization for Poisson, binomial and maybe gamma likelihoods.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Most of the below comes from the [functional-spec](https://github.com/stan-dev/stan/wiki/Functional-Spec:-Sparse-Matrix-Data-Types) for sparse matrices.
+
+A sparse matrix/vector differs from a standard matrix/vector only in the underlying representation of the data.
+
+Below we go over how sparse matrices can be constructed and operated on within each stan block.
+
+
+## Data
+
+Sparse matrix types in the Stan language can be constructed in the data block via Coordinate List Notation using the row and column sizes, non-empty row/column indices, and values for those index positions.
+
+```stan
+data {
+int N; // Rows
+int M; // Cols
+int K; // number non-empty
+int nonzero_row_index[K]; // Non-empty row positions
+int nonzero_col_index[K]; // Non-empty col positions
+vector[N] vals; // Values in each position
+// Direct way
+sparse_matrix[N, M, nonzero_row_index, nonzero_col_index, val] A
+// Can we do this?
+sparse_matrix[N, M, nonzero_row_index, nonzero_col_index] B;
+}
+```
+
+Sparse vectors are the same only with a single size on a single index array.
+
+```stan
+int N; // Rows
+int K; // number non-empty
+int nonzero_row_index[K]; // Non-empty row positions
+vector[N] vals; // Values in each position
+// Direct way
+sparse_matrix[N, nonzero_row_index, val] A
+// Can we do this?
+sparse_vector[N, nonzero_row_index] B;
+```
+
+The above sparse matrix example with values
+
+```stan
+N = 5
+M = 5
+K = 7
+// Column major order
+nonzero_row_index[K] = [2, 3, 1, 3, 5, 3, 2, 5]
+nonzero_col_index[K] = [1, 1, 2, 2, 3, 4, 5, 5]
+val[K] = [22, 7, 3, 5, 14, 1, 17, 8]
+```
+
+Would have the dense form of
+
+| row/col | 1  | 2 | 3  | 4 | 5  |
```

Comment:
Isn't this col/row? ie columns are on top?

Reply:
Yep! Flipped these